src/werkzeug/local.py

```diff
@@ -46,41 +90,78 @@ def release_local(local: Union["LocalStack", "Local"]) -> None:
 
 
 class Local:
-    __slots__ = ("__storage__", "__ident_func__")
+    __slots__ = ("_storage", "_ident_func")
 
     def __init__(self) -> None:
-        object.__setattr__(self, "__storage__", {})
-        object.__setattr__(self, "__ident_func__", get_ident)
+        object.__setattr__(self, "_storage", ContextVar("local_storage"))
+        object.__setattr__(self, "_ident_func", get_ident)
+
+    @property
+    def __storage__(self):
+        warnings.warn(
+            "__storage__ is deprecated", DeprecationWarning, stacklevel=2,
+        )
+        return self._storage.get({})
+
+    @property
+    def __ident_func__(self):
+        if not hasattr(self._storage, "ident_func"):
+            raise RuntimeError(
+                "The __ident_func__ should not be used in Python 3.7+ "
+                "as a ContextVar is used."
+            )
+        else:
+            warnings.warn(
+                "__ident_func__ is deprecated and does not work with Python 3.7+",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            return self._ident_func
+
+    @__ident_func__.setter
+    def __ident_func__(self, func):
+        if not hasattr(self._storage, "ident_func"):
+            raise RuntimeError(
+                "The __ident_func__ cannot be changed in Python 3.7+ "
+                "as a ContextVar is used."
+            )
+        else:
+            warnings.warn(
+                "__ident_func__ is deprecated and does not work with Python 3.7+",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            self._ident_func = func
 
     def __iter__(self):
-        return iter(self.__storage__.items())
+        return iter(self._storage.get({}).items())
```

Comment:
`get({})` is used in a lot of places, and will be creating and throwing away empty dict objects most of the time. Maybe these calls should be extracted to a method that calls it without a default and catches `LookupError`?

Reply:
That's fine, we can always do further refactors later.